### PR TITLE
`cw721-expiration` contract with `invalid NFT` utility

### PR DESCRIPTION
**Pull request by taitruong**
README.md:

# CW721 Expiration

One typical use cases for an `cw721-expiration` contract is providing services for a limited period, like access cards, SLAs, cloud services, etc. Also check kudos below.

This contract enables the creation of NFTs that expire after a predetermined number of days. The `expiration_days` is set during instantiation of contract.
Expiration timestamp is: mint timestamp + expiration days.

### Query Messages

This contract extends cw721-base by adding a new `invalid NFT` utility. The following `cw721-base` query messages have been extended with an optional `include_invalid` property:

- `OwnerOf`: Queries owner of NFT, by default it throws an invalid NFT error.
- `Approval`: Queries whether spender has approval for a NFT, by default it throws an invalid NFT error.
- `Approvals`: Queries all approvals for a NFT, by default it throws an invalid NFT error.
- `NftInfo`: Queries NFT Info data, by default it throws an invalid NFT error.
- `AllNftInfo`: Queries NFT Info data, owner, and approvals, by default it throws an invalid NFT error.
- `Tokens`: Queries all token IDs owned by given address, by default it filters invalid NFTs.
- `AllTokens`: Queries all token IDs, by default it filters invalid NFTs.

In case NFT is invalid (due to expiration) an error is thrown or filtered out. Above queries for including invalid NFTs must explicitly pass `include_invalid: Some(true)` (in all other cases (`None`, `Some(false)`) invalid NFTs are excluded).

### Execute Messages

Execute messages are kept unchanged, but during execution an error is thrown for invalid NFTs for these operations:

- `TransferNft`: Transfers a NFT to another account without triggering an action.
- `SendNft`: Sends a NFT to another account and triggering an action.
- `Approve`: Allows operator/spender to transfer, send, and burn an NFT.
- `Revoke`: Revokes above approval.
- `Burn`: Burns an NFT.

## Instantiation

To instantiate a new instance of this contract you must specify `expiration_days` - along with cw721-based properties: `owner` (aka minter), `name`, and `symbol`.

### CODE REVIEW

**Comment by taitruong** [comment]
Kudos to [timpi](https://timpi.io/) for requesting this kind of NFTs, allowing NFT holders running nodes for a limited period.

**Comment by taitruong** [comment]
Some side notes to be addressed in other tasks. Mainly most of the code was simply __copy-pasted__ from `cw721-base`. Imo this shows `cw721-base` is not modular/generic enough for custom c721-contracts. My suggestions:

- adjust `cw721` contract
  - create a new template (as noted in #118)
  - provide default implementations for traits (see #135)
  - always pass `deps`, `env`, `info` for query and execute messages - otherwise traits cant be used, instead code needs to be duplicated like in this case.
  - other minor changes and renamings like:
    - #136 
    - rename `check_can_send` -> `check_approval`, also approval and operator is kinda confusing
- use Sylvia
  - see PR: https://github.com/CosmWasm/cw-nfts/pull/127
  - also check this on Generics support soon: https://github.com/CosmWasm/sylvia/issues/217

Imo, by doing so, building custom, extended cw721 contracts get much faster done and better, with less/no copy paste code.

**Review by shanev — APPROVED** [review]
LGTM

File: contracts/cw721-expiration/schema/raw/instantiate.json
Comment (JakeHartnell) [inline contracts/cw721-expiration/schema/raw/instantiate.json:1]: Minor nit, can we add the `raw` folder to `.gitignore` so we don't include all these files. Main `cw721-expiration.json` schema is sufficient IMO.

**Review by JakeHartnell — APPROVED** [review]
@taitruong LGTM! Mind fixing the one linting error?

File: contracts/cw721-expiration/schema/raw/instantiate.json
Comment (taitruong) [reply contracts/cw721-expiration/schema/raw/instantiate.json:1]: raw files removed and added to `.gitignore`

**Comment by taitruong** [comment]
> @taitruong LGTM! Mind fixing the one linting error?

Done. Also fixes #141

**Pull request merged**